multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/AbstractKeycloakFilter.java

```diff
@@ -60,4 +84,57 @@ public void init(FilterConfig filterConfig) throws ServletException {}
 
   @Override
   public void destroy() {}
+
+  private class SigningKeyResolverAdapterByKind extends SigningKeyResolverAdapter {
+
+    @Override
+    public Key resolveSigningKey(JwsHeader header, String plaintext) {
+      if (MACHINE_TOKEN_KIND.equals(header.get("kind"))) {
+        throw new JwtException("Not a keycloak token");
+      }
+      try {
+        return getJwtPublicKey(header);
+      } catch (JwkException e) {
+        throw new JwtException(
+            "Error during the retrieval of the public key during JWT token validation", e);
+      }
+    }
+
+    @Override
+    public Key resolveSigningKey(JwsHeader header, Claims claims) {
+      if (MACHINE_TOKEN_KIND.equals(header.get("kind"))) {
+        throw new JwtException("Not a keycloak token");
+      }
+      try {
+        return getJwtPublicKey(header);
+      } catch (JwkException e) {
+        throw new JwtException(
+            "Error during the retrieval of the public key during JWT token validation", e);
+      }
+    }
+  }
+
+  protected synchronized PublicKey getJwtPublicKey(JwsHeader<?> header) throws JwkException {
+    String kid = header.getKeyId();
+    if (kid == null) {
+      LOG.warn(
+          "'kid' is missing in the JWT token header. This is not possible to validate the token with OIDC provider keys");
+      return null;
```

Comment:
Which exception will be thrown if `SigningKeyResolver` returns null `PublicKey`?

Reply:
It will be IllegalArgumentException in the parse mathod

Reply:
I don't see where `IllegalArgumentException` will be catched and processed. Maybe it would be better to throw `JwtException` with meaningful message?